noise/noise.py

```diff
@@ -0,0 +1,157 @@
+    payload = hexlify(struct.pack(
+        "!QQL",
+        int(block) << 40 | int(tx) << 16 | int(out),
+        int(n['amount_msat']),
+        blockheight + n['delay'])).decode('ASCII')
+    payload += "00" * 12
+    return payload
+
+
+def buildpath(plugin, node_id, payload, amt, exclusions):
+    blockheight = plugin.rpc.getinfo()['blockheight']
+    route = plugin.rpc.getroute(node_id, amt, 10, exclude=exclusions)['route']
+    first_hop = route[0]
+    # Need to shift the parameters by one hop
+    hops = []
+    for h, n in zip(route[:-1], route[1:]):
+        # We tell the node h about the parameters to use for n (a.k.a. h + 1)
+        hops.append({
+            "type": "legacy",
+            "pubkey": h['id'],
+            "payload": serialize_payload(n, blockheight)
+        })
+
+    # The last hop has a special payload:
+    hops.append({
+        "type": "tlv",
+        "pubkey": route[-1]['id'],
+        "payload": hexlify(payload).decode('ASCII'),
+    })
+    return first_hop, hops, route
+
+
+def deliver(node_id, payload, amt, max_attempts=5, payment_hash=None):
+    """Do your best to deliver `payload` to `node_id`.
+    """
+    if payment_hash is None:
+        payment_hash = ''.join(random.choice(string.hexdigits) for _ in range(64)).lower()
+
+    exclusions = []
+
+    for attempt in range(max_attempts):
+        logging.debug("Starting attempt {} to deliver message to {}".format(attempt, node_id))
+
+        first_hop, hops, route = buildpath(plugin, node_id, payload, amt, exclusions)
+        onion = plugin.rpc.createonion(hops=hops, assocdata=payment_hash)
+
+        plugin.rpc.sendonion(onion=onion['onion'],
+                             first_hop=first_hop,
+                             payment_hash=payment_hash,
+                             shared_secrets=onion['shared_secrets']
+        )
+        try:
+            plugin.rpc.waitsendpay(payment_hash=payment_hash)
+            return {'route': route, 'payment_hash': payment_hash, 'attempt': attempt}
+        except RpcError as e:
+            print(e)
+            failcode = e.error['data']['failcode']
+            if failcode == 16399:
+                return {'route': route, 'payment_hash': payment_hash, 'attempt': attempt+1}
+
+            plugin.log("Retrying delivery.")
+
+            # TODO Store the failing channel in the exclusions
+    raise ValueError('Could not reach destination {node_id}'.format(node_id=node_id))
+
+
+@plugin.async_method('sendmsg')
+def sendmsg(node_id, msg, plugin, request, amt=1000, **kwargs):
+    payload = BytesIO()
+    varint_encode(34349334, payload)
+    varint_encode(len(msg), payload)
+    payload.write(msg.encode('UTF-8'))
+
+    # Sign the message:
+    sig = plugin.rpc.signmessage(msg)['signature']
+    sig = unhexlify(sig)
+    varint_encode(34349336, payload)
+    varint_encode(len(sig), payload)
+    payload.write(sig)
+
+    res = deliver(node_id, payload.getbuffer(), amt=amt)
```

Comment:
No key send to compensate the routing nodes for their service?

Reply:
I have a local branch that does it, but I needed to have a working branch for the `createonion` and `sendonion` pull request. It's really just a 2 line change to add payments attached to messages :wink: